File: prushka-game/src/main/java/top/nipuru/prushka/game/util/font/Bitmap.java
package top.nipuru.prushka.game.util.font;

public class Bitmap {
    final String[] chars;
    final String string;
    final String font;
    final String name;
    final int width;

    public Bitmap(String name, String font, int width, String... chars) {
        this.name = name;
        this.font = font;
        this.width = width;
        this.chars = chars;
        StringBuilder sb = new StringBuilder();
        for (String s : chars) {
            sb.append(s);
        }
        string = sb.toString();
        if (chars.length == 0) {
            System.out.println("bitmap :" + name + " 为空!");
        }
    }

    public String getName() {
        return name;
    }

    public String getFont() {
        return font;
    }

    public int getWidth() {
        return width;
    }

    public int getSize() {
        return string.length();
    }

    public int getRows() {
        return chars.length;
    }

    public int getRowSize(int row) {
        return chars.length > row ? chars[row].length() : 0;
    }

    public char getChar() {
        return string.charAt(0);
    }

    public char getChar(int index) {
        return string.charAt(index % string.length());
    }

    public char getChar(int row, int col) {
        String str = chars[row % chars.length];
        return str.charAt(col % str.length());
    }

    public String getRange(int startInclude, int endExclude) {
        return string.substring(startInclude % string.length(), Math.min(endExclude, string.length()));
    }

    public String getRange(int row, int startInclude, int endExclude) {
        String str = chars[row % chars.length];
        return str.substring(startInclude % str.length(), Math.min(endExclude, str.length()));
    }
}

File: server/server-bukkit/src/main/kotlin/server/bukkit/util/text/font/Bitmap.kt
package server.bukkit.util.text.font

import net.kyori.adventure.key.Keyed
import kotlin.math.min

class Bitmap(val name: String, val font: Keyed, val width: Int, vararg val chars: String) {

    val string: String

    val size: Int get() = string.length

    val rows: Int get() = chars.size

    init {
        require(chars.isNotEmpty()) { "No characters provided" }
        val sb = StringBuilder()
        for (s in chars) { sb.append(s) }
        string = sb.toString()
    }

    fun getRowSize(row: Int): Int {
        return if (rows > row) chars[row].length else 0
    }

    fun getChar(index: Int = 0): Char {
        return string[index % size]
    }

    fun getChar(row: Int = 0, col: Int = 0): Char {
        val str = chars[row % rows]
        return str[col % str.length]
    }

    fun getRange(startInclude: Int, endExclude: Int): String {
        return string.substring(startInclude % size, min(endExclude, size))
    }

    fun getRange(row: Int, startInclude: Int, endExclude: Int): String {
        val str = chars[row % rows]
        return str.substring(startInclude % str.length, min(endExclude, str.length))
    }
}
